rust/main/agents/relayer/src/msg/pending_message.rs

```diff
@@ -516,44 +517,37 @@ impl PendingOperation for PendingMessage {
 
 impl PendingMessage {
     /// Constructor that tries reading the retry count from the HyperlaneDB in order to recompute the `next_attempt_after`.
+    /// If the message has been retried more than `max_retries`, it will return `None`.
     /// In case of failure, behaves like `Self::new(...)`.
-    pub fn from_persisted_retries(
+    pub fn maybe_from_persisted_retries(
         message: HyperlaneMessage,
         ctx: Arc<MessageContext>,
         app_context: Option<String>,
-    ) -> Self {
-        // Attempt to fetch status about message from database
-        let message_status = match ctx.origin_db.retrieve_status_by_message_id(&message.id()) {
-            Ok(Some(status)) => {
-                // This event is used for E2E tests to ensure message statuses
-                // are being properly loaded from the db
-                tracing::event!(
-                    if cfg!(feature = "test-utils") {
-                        Level::DEBUG
-                    } else {
-                        Level::TRACE
-                    },
-                    ?status,
-                    id=?message.id(),
-                    RETRIEVED_MESSAGE_LOG,
-                );
-                status
-            }
-            _ => {
-                tracing::event!(
-                    if cfg!(feature = "test-utils") {
-                        Level::DEBUG
-                    } else {
-                        Level::TRACE
-                    },
-                    "Message status not found in db"
-                );
-                PendingOperationStatus::FirstPrepareAttempt
-            }
-        };
+        retries_before_skipping: Option<u32>,
+    ) -> Option<Self> {
+        let num_retries = Self::maybe_get_num_retries(
+            ctx.origin_db.clone(),
+            message.clone(),
+            retries_before_skipping,
+        )?;
+
+        let message_status = Self::get_message_status(ctx.origin_db.clone(), message.clone());
+        let mut pending_message = Self::new(message, ctx, message_status, app_context);
+        if num_retries > 0 {
+            let next_attempt_after =
+                PendingMessage::calculate_msg_backoff(num_retries).map(|dur| Instant::now() + dur);
+            pending_message.num_retries = num_retries;
+            pending_message.next_attempt_after = next_attempt_after;
+        }
+        Some(pending_message)
+    }
 
-        let num_retries = match ctx
-            .origin_db
+    fn maybe_get_num_retries(
+        origin_db: Arc<dyn HyperlaneDb>,
+        message: HyperlaneMessage,
```

Comment:
This method does not need to take ownership of `message`. I wonder if passing as a reference will be sufficient.